modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java

```diff
@@ -17,213 +17,376 @@
      * TODO Documentation https://issues.apache.org/jira/browse/IGNITE-15859
      */
-    public synchronized void onTableDropped(
+    public CompletableFuture<Void> onTableDropped(
             String schemaName,
-            String tableName
+            String tableName,
+            long causalityToken
     ) {
-        IgniteSchema schema = igniteSchemas.computeIfAbsent(schemaName, IgniteSchema::new);
-
-        InternalIgniteTable table = (InternalIgniteTable) schema.getTable(tableName);
-        if (table != null) {
-            tablesById.remove(table.id());
-            schema.removeTable(tableName);
-        }
-
-        rebuild();
+        final AtomicReference<InternalIgniteTable> tableRef = new AtomicReference<>();
+
+        return schemasVv
+                .update(causalityToken,
+                    schemas -> {
+                        IgniteSchema prevSchema = schemas.computeIfAbsent(schemaName, IgniteSchema::new);
+                        IgniteSchema schema = new IgniteSchema(prevSchema.getName(), prevSchema.getTableMap());
+
+                        InternalIgniteTable table = (InternalIgniteTable) schema.getTable(tableName);
+
+                        if (table != null) {
+                            schema.removeTable(tableName);
+                        }
+
+                        tableRef.set(table);
+
+                        return PatchedMapView.of(schemas).put(schemaName, schema);
+                    },
+                    e -> {
+                        throw new IgniteInternalException(e);
+                    }
+                )
+                .thenApply(s -> tableRef.get())
+                .thenCompose(table ->
+                    tablesVv.update(causalityToken,
+                        tables -> PatchedMapView.of(tables).remove(table.id()),
+                        e -> {
+                            throw new IgniteInternalException(e);
+                        }
+                    )
+                )
+                .thenCompose(t -> {
+                    try {
+                        return rebuild(causalityToken);
+                    } catch (OutdatedTokenException e) {
+                        throw new IgniteInternalException(e);
+                    }
+                });
     }
 
-    private void rebuild() {
+    private CompletableFuture<Void> rebuild(long causalityToken) throws OutdatedTokenException {
         SchemaPlus newCalciteSchema = Frameworks.createRootSchema(false);
 
         newCalciteSchema.add("PUBLIC", new IgniteSchema("PUBLIC"));
 
-        igniteSchemas.forEach(newCalciteSchema::add);
-        externalCatalogs.forEach(newCalciteSchema::add);
-
-        calciteSchema = newCalciteSchema;
+        return CompletableFuture.allOf(
+            schemasVv.get(causalityToken).thenAccept(schemas -> schemas.forEach(newCalciteSchema::add)),
+            externalCatalogsVv.get(causalityToken).thenAccept(catalogs -> catalogs.forEach(newCalciteSchema::add))
```

Comment:
I think, here previous token is required (previous from the causalityToken).

Reply:
rewrote it, with VersionedValue#update returning T it became more simple.